tests/test.sh

```diff
@@ -0,0 +1,164 @@
+    else
+      $ODO_PATH create "$devfileName" --starter || error=true
+    fi
+
+    if $error; then
+        echo "ERROR create failed"
+        $ODO_PATH project delete -f "$devfileName"
+        FAILED_TESTS="$FAILED_TESTS $devfileName"
+        return 1
+    fi
+
+    if [ "$ENV" = "minikube" ]; then
+        # ToDo: Clean up, I'm not happy about having specific checks for the stacks with multiple ports
+        # But since we're testing against minikube, we need to specifically create the URL/ingress before pushing
+        # And if there's multiple ports in the devfile, a port must be specified.
+        if [ "$devfileName" = "java-wildfly" ] || [ "$devfileName" = "java-wildfly-bootable-jar" ]; then
+            $ODO_PATH url create --host "$(minikube ip).nip.io" --port 8080 || error=true
+            $ODO_PATH url create --host "$(minikube ip).nip.io" --port 16686 || error=true
+        else
+            $ODO_PATH url create --host "$(minikube ip).nip.io" || error=true
+        fi
+        if $error; then
+            echo "ERROR url create failed"
+            $ODO_PATH project delete -f "$devfileName"
+            FAILED_TESTS="$FAILED_TESTS $devfileName"
+            return 1
+        fi
+    fi
+
+    $ODO_PATH push || error=true
+    if $error; then
+        echo "ERROR push failed"
+        $ODO_PATH project delete -f "$devfileName"
+        FAILED_TESTS="$FAILED_TESTS $devfileName"
+        return 1
+    fi
+
+    # check if application is responding
+    urls=$(getURLs)
+
+    for url in $urls; do
+        statusCode=200
+        
+        waitForHTTPStatus "$url" "$statusCode"
+        if [ $? -ne 0 ]; then
+            echo "ERROR unable to get working url"
+            $ODO_PATH project delete -f "$devfileName"
+            FAILED_TESTS="$FAILED_TESTS $devfileName"
+            error=true
+            return 1
+        fi
+    done
+
+    # kill -9 $CPID
+    $ODO_PATH delete -f -a || error=true
+    $ODO_PATH project delete -f "$devfileName"
+
+    if $error; then
+        echo "FAIL"
+        # record failed test
+        FAILED_TESTS="$FAILED_TESTS $devfileName"
+        return 1
+    fi
+
+    return 0
+}
+
+
+ODO_PATH=$1
+if [ -z $ODO_PATH ]; then
+  ODO_PATH=odo
+fi
+if [ -z $ENV ]; then
+  ENV=minikube
+fi
+if [ "$ENV" != "minikube"] && ["$ENV" != "openshift" ]; then
+  echo "ERROR: Allowed values for ENV are either \"minikube\" (default) or \"openshift\"."
+  exit 1
+fi
+if [ -z $REGISTRY ]; then
+  REGISTRY=local
+fi
+if [ "$REGISTRY" != "local" ] && ["$REGISTRY" != "remote" ]; then
+  echo "ERROR: Allowed values for REGISTRY are either \"local\" (default) or \"remote\"."
+  exit 1
+fi
+
+for devfile_dir in $(find $DEVFILES_DIR -maxdepth 1 -type d ! -path $DEVFILES_DIR); do
+    devfile_name="$(basename $devfile_dir)"
+    devfile_path=$devfile_dir/devfile.yaml
+    if [ $devfile_name != "java-wildfly-bootable-jar" ]; then
+      test "$devfile_name" "$devfile_path"
```

Comment:
I think we need to check the return code of test() here. Then if the test() fails we can either exit the test script immediately or record the failure and return all the results at the end of the test script.

Reply:
I'd say let's leave it as-is for now, since it works fine either way. We can always revisit later if needed.